### PR TITLE
Add Chow-Liu and HMM image region graphs

### PR DESCRIPTION
**Pull request by IrwinChay**
This PR does: 
1. enabled the use of CLT `chow-liu-tree` and HMM `linear-tree` region graphs in constructing region graph from image data
2. added chunked mutual information (over data axis) for Chow-Liu Tree with Gaussian inputs 
3. Corresponding tests for both 

Notice that:  
1. In `chow_liu.py`, in chunked mutual information for CLT, the implementation in Monarch [Circuit](https://github.com/wangben88/MonarchCircuits/blob/master/monarch_juice/structures/HCLTMonarch.py) chunked over the feature axis, while our existing categorical case ( `_categorical_mutual_info` ) chunked over the data axis. For categorical inputs, it is more memory efficient to chunk over the feature axis (O(feature*samples) vs O(feature^2 * categories^2)); but for the gaussian inputs, it is bette to chunk over the data axis (O(feature*samples)) difference. However, the images we use won't be extremely large so both options could fit into memory. So for consistency, chunked mutual information along the data axis is implemented for Gaussian inputs. 
2. In `data_modalities.py`, note that the existing `LinearTree` region graph is just HMM region graph, I simply flattens the image to linear order since other [code base](https://github.com/Tractables/pyjuice/blob/09e20a5539c2602d6ffee7b9119cb083a711a4d0/src/pyjuice/structures/hmm.py) does the same

### CODE REVIEW

File: cirkit/backend/torch/rules/layers.py [hunk 1]
```diff
@@ -37,9 +37,7 @@
     from cirkit.backend.torch.compiler import TorchCompiler
 
 
-def compile_embedding_layer(
```
Comment (chbricout) [inline cirkit/backend/torch/rules/layers.py]: I think you might be formatting with a different tool, try running black on the code to avoid having this diffs please!

File: tests/data_modalities/test_image_data.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        data=data,
+        num_bins=num_bins,
+        mi_chunk_size=mi_chunk_size,
+        sum_weight_param=utils.Parameterization(activation="softmax", initialization="normal"),
+    )
+
+    # Check if the finite log-likelihoods has the expected shape
+    assert len(symbolic_circuit.scope) == num_variables
+    circuit = compile(symbolic_circuit)
+    ll = circuit(data)
```
Comment (chbricout) [inline tests/data_modalities/test_image_data.py]: Could you name this `output`, the circuit is not giving a log likelihood when using embedding

File: tests/templates/region_graph/test_algorithms.py [hunk 3]
```diff
@@ -144,3 +147,53 @@ def test_rg_algorithm_poon_domingos(
...
+        discrete = torch.randint(0, 4, (num_samples, num_variables - num_continuous)).float()
+        data = torch.cat([continuous, discrete], dim=1)
+    return data
+
+
+@pytest.mark.parametrize(
+    "input_type,chunk_size",
+    itertools.product(["categorical", "gaussian", "heterogeneous"], [None, 256]),
+)
+def test_rg_algorithm_chow_liu_tree(input_type: str, chunk_size: int | None):
```
Comment (chbricout) [inline tests/templates/region_graph/test_algorithms.py]: Could it be possible to check the output of the algorithm against a manually computed CLT on toy data?

File: cirkit/templates/region_graph/algorithms/chow_liu.py
Comment (chbricout) [inline cirkit/templates/region_graph/algorithms/chow_liu.py:19]: Could you tell me if you have an equivalent to the `bin_for_mi` argument implemented in this repo ? https://github.com/dylanponsford/cirkit/blob/ae148408dfdce5a01e018d401476ed0909ec0f11/cirkit/templates/region_graph/algorithms/chow_liu.py#L19

It was introduced on this fork and we don't have it in cirkit yet.

**Comment by chbricout** [comment]
Good PR ! I have left a few comments, tell me if you have any questions.

**Review by chbricout — CHANGES_REQUESTED** [review]

File: cirkit/backend/torch/rules/layers.py (hunk 1 above)
Comment (IrwinChay) [reply cirkit/backend/torch/rules/layers.py]: I saw you fixed this in your last commit

File: tests/data_modalities/test_image_data.py (hunk 2 above)
Comment (IrwinChay) [reply tests/data_modalities/test_image_data.py]: Done

File: tests/templates/region_graph/test_algorithms.py (hunk 3 above)
Comment (IrwinChay) [reply tests/templates/region_graph/test_algorithms.py]: Done

File: cirkit/templates/region_graph/algorithms/chow_liu.py
Comment (IrwinChay) [reply cirkit/templates/region_graph/algorithms/chow_liu.py:19]: Added. Also added its docstrings and tests

Comment (chbricout) [reply cirkit/templates/region_graph/algorithms/chow_liu.py:19]: Nice ! Thank you

**Comment by chbricout** [comment]
Everything looks good to me, before we merge I was just wondering if it would not make sense to rename `num_bins` and `bin_for_mi` in somethin like `cat_bins` and `cont_bins` so it's easier to understand what they are doing. What do you think?

Any opinions @arranger1044 @dylanponsford (as people that are working with cirkit).

**Comment by IrwinChay** [comment]
> Everything looks good to me, before we merge I was just wondering if it would not make sense to rename `num_bins` and `bin_for_mi` in somethin like `cat_bins` and `cont_bins` so it's easier to understand what they are doing. What do you think?
> 
> Any opinions @arranger1044 @dylanponsford (as people that are working with cirkit).

Good idea. Changed to `cat_bins` and `heter_cont_bins` because the it is only for heterogeneous continuous inputs

**Comment by chbricout** [comment]
@IrwinChay I was not expecting you to be this efficient! I wanted to have a discussion about the variable names, I think the one you chose are very good though. One thing that still needed to be discuss is whether we should keep `num_bins` as a deprecated argument to keep backward compatibility with 0.3.x. Maybe we can plan to remove it in 0.4.0. Let's see what @arranger1044 thinks before doing it!
